Release process (#3379)
input-output-hk/cardano-ledger: RELEASING.md

repository are build together at the same time should be ignored. It should be assumed
that every package can be released to CHaPs by itself at any point in time. This is
necessary to guarantee that if such a package is compatible with other packages from this
repository, which have been previously released, the version bounds should reflect that.

### Upper bounds

Normally, [strict upper bounds are required by
PVP](https://pvp.haskell.org/#dependencies-in-cabal) for all dependencies of a
package. This is the only exception we make to the PVP, and we do not require strict upper
bounds for dependenicies that come from Hackage. For exmaple there is no need to add an
upper bound on `containers` package, because that's just dumb. On the other hand all of
the dependeices that come from CHaPs should follow strict bounds, because those packages
are a lot more volatile and creating revisions for packages on CHaPs is not as smooth as
it is on Hackage.

### `CHANGELOG.md`

Every package, that [is subject to the release
process](packages-excluded-from-release-process), will always have the top most entry in

> the `CHANGELOG.md` set to an unreleased version. That top most section will have either

Reviewer: Did we agree to putting "(unreleased)" next to the top entry?

For example:

```
1.2.0 (unreleased)
* stuff

1.1.0
* things
```

This way, when `1.2.0` is released, we do not need to do something like add a new `1.2.1` or some such as a part of release process (to maintain the invariant that we "always have the top most entry in the `CHANGELOG.md` set to an unreleased version").

Reviewer: chatted in a call, I like the new plan!

the minor or the major versions bumped, but not both, when compared to the latest released
version. Which one it will be depends on the recent changes that were added after the
latest release was made. More on why this should always be true is in the [Release
Process](#release-process) section. For example, if the latest version of
`cardano-ledger-core` that was released to CHaPs is `1.20.1.0`, then there will be a
section like this:

```
# Version history for `cardano-ledger-core`

## 1.20.2.0

* Add ...
...
```

> It is quite common to experience conflicts in the changelog, since that will be the most

Reviewer 2: You might enjoy using `scriv`, we've been using it fairly happily in `plutus` to solve this.

Author: Nah, `scriv` is really a hammer that we don't need. It would be very useful if we were dealing with bundles as Plutus or Consensus does, but we don't have this problem. Moreover, we as developers, deal with git conflicts on daily basis, so couple more conflicts in the changelog is not a big deal.

common section of the codebase being updated at the same time. With respect on conflicts
resolution on the package version it is pretty easy, the highest value should usually win.

With respect to the actual change log entries, there is no enforced order in which entries
should be added. To the top, to the bottom or in the middle of the list, it is
irrelevant. In fact, randomized insertion reduces conflicts, so feel free to add entries
anywhere in the section. Most important is that we can always infer the actual order from
the git history, which provide with a very reliable source of order on changes.

#### When to update

> `CHANGELOG.md` must be updated for the package that is expereincing a change and any other

Reviewer 3: ```suggestion
`CHANGELOG.md` must be updated for the package that is experiencing a change and any other
```

package in the repository that re-exports that change in one form or another. In general
an entry is added according to these rules:

* Every breaking change MUST appear in the `CHANGELOG.md`.

* Every non-breaking change SHOULD appear in the `CHANGELOG.md`

* Every change that does not change the user facing semantics of the code (eg. changes to
  documentation, depenency bounds, test suites, addition of internal functions,
  performance improvements, etc.)  COULD be added to the `CHANGELOG.md`, but they are
  discouraged. Exceptions should be made for changes that could be valueable to the
  downstream user.

* Every change to a public sub-library of a package must be added to a separate section in
  the `CHANGELOG.md` and is versioned together with the main library. Eg. `testlib`:

  ```
  # Version history for `cardano-ledger-core`

  ## 1.20.2.0

  ### `testlib`

  * Add `Arbitrary` instance for ...
  ...
  ```

## Release Process

> Usually a package release will happen directly from `master` branch. Current release

Reviewer: We should remove mention of the "trunk based developement" then: https://github.com/input-output-hk/cardano-ledger/blob/master/CONTRIBUTING.md#branching-model

Author: I've removed the link and other no longer relevant information. I have left a mention about the trunk based development, since technically that is still our model. We just do releasing differently, since we have many packages that we can release, not just one.

Reviewer: ok, thanks!

process even allows for a release to happen from a commit that is slightly behind `HEAD`
of `master` branch, but every major release must happen from `master`. On a rare occasion
when a bugfixes need to be backported to an older version, a release can happen from an
ephemeral release branch. Such a release branch must follow the same procedure of a PR
review and CI, that is why it should be prefixed with `release/some-branch-name`, cause

> taht will trigger the CI. should be followed during a bugfix release to an older version,

Reviewer 4: ```suggestion
that will trigger the CI. should be followed during a bugfix release to an older version,
```
Also seems like there's something missing from the beginning of the sentence.

Author: Fixed. Thanks

that has diverged from master, as for the PR into master directly.

### Release Steps

> Here are the steps for release engineer to folow.

Reviewer 4: ```suggestion
Here are the steps for release engineer to follow.
```

#### Decide what to release

Normally it should be OK to release at the same time all packages from the repository that
experienced some sort of change, but that is not mandatory. Any package can be released by
itself at any point from `HEAD` of `master` or a even commit that is behind the `HEAD` of
`master`. A release with minor version bump from an ephemeral release branch is also
possible.

The most common case is to blindly release all of the packages that were changed since

> they werre last released. This begs a question. How to decide which of the packaged have

Reviewer 4: ```suggestion
they were last released. This begs a question. How to decide which of the packages have
```

changed, thus deserve a release?

The rule is very simple. Every package that falls under this release process and has a
version in its cabal file higher than the highest version released to CHaPs is allowed to

> be released. It sis possibly to rely on git tags for information about the latest released

Reviewer 4: ```suggestion
be released. It is possible to rely on git tags for information about the latest released
```

   `1.20.2.0` section in the `CHANGELOG.md` must be added:

   ```markdown
   # Version history for `cardano-ledger-core`

   ## 1.20.2.0

   *

   ## 1.20.1.1

   ...
   ```

#### Backporting changes

It is occasionally necessary to release a minor version for a package that has its history
diverged from a version on `master` significantly enough to make it impossible for a minor
version to be released from `master`. In other words a patch backporting. In such a
scenario a few steps should be followed:

1. A branch with a prefix `release/` need to be created from a tag of a package version
   that is beeing updated. For example if a current version on `master` is
   `cardano-ledger-core-1.22.10.0` then the latest `cardano-ledger-core-1.21.x` should be
   used as base:

   ```shell
   $ git checkout -b release/cardano-ledger-core-1.21.3.0 cardano-ledger-core-1.21.2.1
   ```

2. Changes that need to be released should be `cherry-pick`ed from master. If a fix on
   `master` was implemented in some incompatible fashion to the current release, then it
   is fine to reimplement it anew, as long as the change being introduced is also present

>    on master in some form. That also includes a changelog entry, it should be presetn in

Reviewer 4: ```suggestion
   on master in some form. That also includes a changelog entry, it should be present in
```

   both the patched version and in the next version released from `master`.

3. Regular release process should follow from here.

4. Once the package has been released and a git tag for that release was created, the
   `release/` branch can be removed.

This process does not accomodate backporting fixes to versions that are at least two major
versions behind the one on `master`.

> ### Deciding what to realease

Reviewer 3: ```suggestion
### Deciding what to release
```

> Figure out what's dirty is hard and should be improved

Reviewer: ```suggestion
Figure out what's dirty is hard and should be improved.
```

## Packages excluded from release process

We release most of the packages in this repo to [CHaP (Cardano Haskell
Packages)](https://github.com/input-output-hk/cardano-haskell-packages). However, there
are a few packages that are either used for testing, debugging or benchmarking and do not
deserve to be released into the World. They are neither released nor versioned. Bounds on
the local depndencies do not need to be updated, because they will always use the versions

> for ledger packages from whith the repositiory, rather than from CHaPs. Here is the full

Reviewer 4: ```suggestion
for ledger packages from within the repositiory, rather than from CHaPs. Here is the full
```

# Versioning and release processes

## Versioning Process

Ledger team has fully adopted the [PVP versioning scheme](https://pvp.haskell.org/) for
all packages in this repo.

Updating the version for every package in this repository is the responsibility of the
developer that is introducing a change to any package that is subject to release
process. Version must be updated in both the `CHANGELOG.md` and in the `.cabal` file for
the package in the same PR, as well as every other package that is affected by the change
in the repository. Every affected package that is subject to release process must receive
a lower and, potentially, an upper bound update on the dependency that has experienced the
change.

For example if I add a function to `cardano-ledger-core` and use it in
`cardano-ledger-conway`, then I must add a minor version bump to `cardano-ledger-core` and
I also must add a lower version restriction in `cardano-ledger-conway` for
`cardano-ledger-core`, as per PVP.

In other words, when it comes to versioning, the fact that all of the packages in this

> repository are build together at the same time should be ignored. It should be assumed

Reviewer 5: Just a little nitpick:  `are built together`

Author: That's not a nitpick but a legit spelling mistake. :slightly_smiling_face: Thank you.

that every package and all of its required dependencies can be released to CHaPs at any
point in time. This is necessary to guarantee that if such a package is compatible with
other packages from this repository, which have been previously released, the version
bounds should reflect that.

### Upper bounds

Normally, [upper bounds are required by
PVP](https://pvp.haskell.org/#dependencies-in-cabal) for all dependencies of a
package. This is the only exception we make to the PVP, and we do not require strict upper
bounds for dependencies that come from Hackage. For example, there is no need to add an
upper bound on `containers` package, unless there is a known incompatibility. Speculative
upper bounds over time lead to overly restrictive build plans that are forced to favor
older packages. Moreover, they lead to a lot of redundant maintenance burden. That being
said, most of the packages from within the same repository are a lot more volatile and

> require tight interoperability, therefore speculative upper bounds are encouraged.

Reviewer 6: We are taking about the ledger-state repository when it says "within the same repository" right?
So might it be better to say,  we encourage upper bounds on packages within the ledger-state repository.

Author: Yeah, you are right, I'll reword it to be more ledger specific.

Author: Done

### Lower bounds

> Placing lower bounds is a lot less controversial than upper bounds, because the don't

Reviewer 5: `they` don't suffer

suffer from speculation about the future. Older versions of dependencies are known at
release time. Therefore lower bounds must be added for all dependencies that are known to

> be incompatible with the package

Reviewer 5: So, say we make  change in `cardano-core` , which `cardano-ledger-babbage` depends on, but this change is not directly impacting `cardano-ledger-babbage` ( so nothing has to be changed in the `cardano-ledger-babbage` to acccomodate the change). 
In this case we leave the lower bound of `cardano-ledger-babbage` as it is, correct?

Author: That is absolutely correct.
